File: engine/src/com/magma/engine/chars/CharacterParty.java
package com.magma.engine.chars;

import java.util.Iterator;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.math.GridPoint2;
import com.badlogic.gdx.scenes.scene2d.Action;
import com.badlogic.gdx.utils.Queue;
import com.magma.engine.debug.MagmaLogger;

public class CharacterParty extends Action {

    private final Character leader;
    private final Queue<CharacterMover> members;
    private final Queue<GridPoint2> trail;

    public CharacterParty(Character leader) {
        this.leader = leader;
        this.trail = new Queue<GridPoint2>();
        this.members = new Queue<CharacterMover>();

        leader.addAction(this);
    }

    public void join(Character newChar) {
        // inject characterMover
        CharacterMover mover = new CharacterMover(newChar.getEntity());
        newChar.addAction(mover);
        members.addLast(mover);
        Gdx.app.log("CharacterParty", newChar.getClass().getSimpleName() + " joined the party!");
    }

    public void joinFirst(Character newChar) {
        // inject characterMover
        CharacterMover mover = new CharacterMover(newChar.getEntity());
        newChar.addAction(mover);
        members.addFirst(mover);
        Gdx.app.log("CharacterParty", newChar.getClass().getSimpleName() + " joined the party in front!");
    }

    public void leave(CharacterMover mover) {
        MagmaLogger.log(this, mover.getActor().getClass().getSimpleName() + " left the party!");
        members.removeValue(mover, false);
        mover.stop();
    }

    public void leave(Character curChar) {
        Iterator<CharacterMover> it = members.iterator();
        while (it.hasNext()) {
            CharacterMover mover = it.next();
            if (mover.getActor() == curChar) {
                break;
            }
        }
    }

    public void disband() {
        Iterator<CharacterMover> it = members.iterator();
        while (it.hasNext()) {
            CharacterMover mover = it.next();
            leave(mover);
        }
    }

    @Override
    public boolean act(float delta) {
        // check if moved
        int x = (int) (getActor().getX());
        int y = (int) (getActor().getY());

        while (trail.size < members.size) {
            trail.addLast(new GridPoint2());
        }

        GridPoint2 head = trail.first();
        if (head.x != x || head.y != y) {
            // player moved
            trail.addFirst(new GridPoint2(x, y));

            // move all members
            for (int i = 0; i < trail.size - 1; i++) {
                GridPoint2 pos = trail.get(i);
                members.get(i).target(pos);
            }

        }

        while (trail.size > members.size) {
            trail.removeLast();
        }

        return false;
    }

    @Override
    public String toString() {
        String text = "CharacterParty: ";
        if (trail.isEmpty()) {
            text += "(no pos)";
            return text;
        }

        for (int i = 0; i < trail.size; i++) {
            GridPoint2 pos = trail.get(i);
            text += " > " + pos.toString();
            if (i < members.size) {
                text += " " + members.get(i).getActor().getClass().getSimpleName();
            }
        }
        return text;
    }
}

File: engine/src/com/magma/engine/chars/CharacterParty.kt
package com.magma.engine.chars

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.math.GridPoint2
import com.badlogic.gdx.scenes.scene2d.Action
import com.badlogic.gdx.utils.Queue
import com.magma.engine.debug.MagmaLogger

class CharacterParty(leader: Character) : Action() {
    private val members: Queue<CharacterMover> = Queue()
    private val trail: Queue<GridPoint2> = Queue()

    init {
        leader.addAction(this)
    }

    fun join(newChar: Character) {
        // inject characterMover
        val mover = CharacterMover(newChar.entity)
        newChar.addAction(mover)
        members.addLast(mover)
        Gdx.app.log("CharacterParty", newChar.javaClass.simpleName + " joined the party!")
    }

    fun joinFirst(newChar: Character) {
        // inject characterMover
        val mover = CharacterMover(newChar.entity)
        newChar.addAction(mover)
        members.addFirst(mover)
        Gdx.app.log("CharacterParty", newChar.javaClass.simpleName + " joined the party in front!")
    }

    fun leave(mover: CharacterMover) {
        MagmaLogger.log(this, mover.actor.javaClass.simpleName + " left the party!")
        members.removeValue(mover, false)
        mover.stop()
    }

    fun leave(curChar: Character) {
        val it: Iterator<CharacterMover> = members.iterator()
        while (it.hasNext()) {
            val mover = it.next()
            if (mover.actor === curChar) {
                break
            }
        }
    }

    /**
     *  Everyone in the party leaves
     */
    fun disband() {
        val it: Iterator<CharacterMover> = members.iterator()
        while (it.hasNext()) {
            val mover = it.next()
            leave(mover)
        }
    }

    override fun act(delta: Float): Boolean {
        // check if moved
        val x = getActor().x.toInt()
        val y = getActor().y.toInt()
        while (trail.size < members.size) {
            trail.addLast(GridPoint2())
        }
        val head = trail.first()
        if (head.x != x || head.y != y) {
            // player moved
            trail.addFirst(GridPoint2(x, y))

            // move all members
            for (i in 0 until trail.size - 1) {
                val pos = trail[i]
                members[i].target(pos)
            }
        }
        while (trail.size > members.size) {
            trail.removeLast()
        }
        return false
    }

    override fun toString(): String {
        var text = "CharacterParty: "
        if (trail.isEmpty) {
            text += "(no pos)"
            return text
        }
        for (i in 0 until trail.size) {
            val pos = trail[i]
            text += " > $pos"
            if (i < members.size) {
                text += " " + members[i].actor.javaClass.simpleName
            }
        }
        return text
    }
}